platforms/software/publish/src/main/java/org/gradle/api/publish/internal/metadata/ModuleMetadataSpecBuilder.java

```diff
@@ -309,6 +314,41 @@ private List<ModuleMetadataSpec.Capability> capabilitiesFor(Collection<? extends
         return metadataCapabilities;
     }
 
+    private static List<ModuleMetadataSpec.Capability> capabilitySelectorsFor(
+        Set<CapabilitySelector> capabilitySelectors,
+        ModuleMetadataSpec.DependencyCoordinates targetComponent
+    ) {
+        if (capabilitySelectors.isEmpty()) {
+            return emptyList();
+        }
+
+        ArrayList<ModuleMetadataSpec.Capability> metadataCapabilities = new ArrayList<>();
+        for (CapabilitySelector capabilitySelector : capabilitySelectors) {
+            if (capabilitySelector instanceof SpecificCapabilitySelector) {
+                SpecificCapabilitySelector specificSelector = (SpecificCapabilitySelector) capabilitySelector;
+                metadataCapabilities.add(
+                    new ModuleMetadataSpec.Capability(
+                        specificSelector.getGroup(),
+                        specificSelector.getName(),
+                        null
+                    )
+                );
+            } else if (capabilitySelector instanceof FeatureCapabilitySelector) {
+                FeatureCapabilitySelector featureSelector = (FeatureCapabilitySelector) capabilitySelector;
+                metadataCapabilities.add(
+                    new ModuleMetadataSpec.Capability(
+                        targetComponent.group,
+                        targetComponent.name + "-" + featureSelector.getFeatureName(),
+                        null
+                    )
+                );
+            } else {
+                throw new UnsupportedOperationException("Unsupported capability selector type: " + capabilitySelector.getClass().getName());
+            }
```

Comment:
💅 Perhaps worth making a factory method here:
```
public static ModuleMetadataSpec.Capability from(CapabilitySelector selector)
```
and hiding these details in that type.

Reply:
Pulled this out into:
```
private static ModuleMetadataSpec.Capability resolveCapability(
        ModuleMetadataSpec.DependencyCoordinates targetComponent,
        CapabilitySelector capabilitySelector
) {
    // ...
}